[FSDP] Delay moving tensor to CPU until necessary for optim_state_dict()

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85761
* #85740

Optimizer state_dict currently move tensors to CPU() immediately after allgather(). However, for sharded optimizer state_dict, this moving is duplicated. We should wait until all the sharding are done. This PR may slightly reduce the performance of full optimizer state_dict as it has to allocate more memory than w/o this PR. But the benchmark shows the memory allocation is pretty light.

Differential Revision: [D39855912](https://our.internmc.facebook.com/intern/diff/D39855912/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39855912/)!

fegin committed:
[FSDP] Delay moving tensor to CPU until necessary for optim_state_dict()

Optimizer state_dict currently move tensors to CPU() immediately after allgather(). However, for sharded optimizer state_dict, this moving is duplicated. We should wait until all the sharding are done. This PR may slightly reduce the performance of full optimizer state_dict as it has to allocate more memory than w/o this PR. But the benchmark shows the memory allocation is pretty light.

Differential Revision: [D39855912](https://our.internmc.facebook.com/intern/diff/D39855912/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39855912/)!

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -132,6 +132,12 @@ def _unflatten_optim_state(
     )
+    if to_save:
+        for optim_state in unflat_param_state:
+            for key in list(optim_state.keys()):
```
Same as the previous PR: I wonder why we do `list(optim_state.keys())`?

fegin replied:
If we use optim_state.items() and modify the dictionary in-place, python will complain. Even though using keys() won't have this issue, I am used to make it a list to avoid the confusion (not iterating and modifying the same data structure). I can remove this if that causes confusion.

awgu replied:
No worries. I was just curious.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -181,25 +186,24 @@ def _communicate_optim_state(
+            # Assume that positive-dimension tensor optimizer state
+            # has the same shape as the sharded flattened parameter
+            buffer_size = flat_param._full_param_padded.size()  # type: ignore[attr-defined]
+            tensor_buffer = value.new_zeros(*buffer_size)
```
Curious why we remove the reuse of `tensor_buffer`? For Adam, this allocates `tensor_buffer` twice per parameter instead of once?

fegin replied:
We can reuse the tensor_buffer because the original code move it to CPU immediately. This PR delay cpu() call, so we have to allocate the buffer twice. It does not add too much cost for the model I tested.

fegin committed:
Update on "[FSDP] Delay moving tensor to CPU until necessary for optim_state_dict()"


Optimizer state_dict currently move tensors to CPU() immediately after allgather(). However, for sharded optimizer state_dict, this moving is duplicated. We should wait until all the sharding are done. This PR may slightly reduce the performance of full optimizer state_dict as it has to allocate more memory than w/o this PR. But the benchmark shows the memory allocation is pretty light.

Differential Revision: [D39855912](https://our.internmc.facebook.com/intern/diff/D39855912/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39855912/)!

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Delay moving tensor to CPU until necessary for optim_state_dict()"


Optimizer state_dict currently move tensors to CPU() immediately after allgather(). However, for sharded optimizer state_dict, this moving is duplicated. We should wait until all the sharding are done. This PR may slightly reduce the performance of full optimizer state_dict as it has to allocate more memory than w/o this PR. But the benchmark shows the memory allocation is pretty light.

Differential Revision: [D39855912](https://our.internmc.facebook.com/intern/diff/D39855912/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D39855912/)!

[ghstack-poisoned]

rohan-varma reviewed (approved):
thanks!
